cassandra/src/main/java/org/hibernate/ogm/datastore/cassandra/CassandraDialect.java

```diff
@@ -210,12 +236,61 @@ public void removeTuple(EntityKey key, TupleContext tupleContext) {
 
 	@Override
 	public Association getAssociation(AssociationKey key, AssociationContext associationContext) {
-		return null; // TODO
+
+		String table = key.getTable();
+		Table tableMetadata = provider.getMetaDataCache().get( key.getTable() );
+		List<Column> tablePKCols = (List<Column>) tableMetadata.getPrimaryKey().getColumns();
+
+		Select select = select().all().from( quote( key.getTable() ) );
+		Select.Where selectWhere = select.where( eq( quote( key.getColumnNames()[0] ), QueryBuilder.bindMarker() ) );
+		for ( int i = 1; i < key.getColumnNames().length; i++ ) {
+			selectWhere = selectWhere.and( eq( quote( key.getColumnNames()[i] ), QueryBuilder.bindMarker() ) );
+		}
+
+		boolean requiredFiltering = false;
+		for ( Column column : tablePKCols ) {
+			String name = column.getName();
+			boolean foundColumn = false;
+			for ( int i = 0; i < key.getColumnNames().length; i++ ) {
+				if ( name.equals( key.getColumnNames()[i] ) ) {
+					foundColumn = true;
+					break;
+				}
+			}
+			if ( !foundColumn ) {
+				requiredFiltering = true;
+				break;
+			}
+		}
+
+		if ( requiredFiltering ) {
+			select.allowFiltering();
+		}
+
+		Object[] columnValues = key.getColumnValues();
+		ResultSet resultSet = bindAndExecute( columnValues, select );
+
+		if ( resultSet.isExhausted() ) {
+			return null;
+		}
+
+		Association association = new Association(
+				new ResultSetAssociationSnapshot(
+						key,
+						resultSet,
+						provider.getMetaDataCache()
```

Comment:
IIRc it stems from the fact that the effective storage primary key for rows in some association tables is the combined key of the parent and the association. Unlike a relational db, cassandra won't allow access by anything other than the primary key first *in order from the first*, i.e. it's prefix matching, so the association key by itself is not enough in cases where the parent key comes first. Ties back into to the comment above about filtering and denormalization - if we e.g. denormalized into 2 tables, one for each side, we'd be able to pick the one with the primary key fields in the right order. Doesn't solve the problem of keeping the other one in sync though, unless we read one to obtain the fields needed to key into the other, which is going to be slow. Slinging additional meatdata around is probably still going to be useful.